### PR TITLE
Package artifacts for all OS and README update

### PR DESCRIPTION
**Pull request by shhsu**

### CODE REVIEW

**Comment by shhsu** [comment]
@sajaya @davidobando

File: README.md [hunk 1]
```diff
@@ -1,2 +1,28 @@
-This is a wrapper for Docker Credential helpers created by Azure Container Registry (ACR) team. This tool allows Azure Active Directory (AAD) based login. The credentials will be saved by the underlying cred-helpers configured at build time.
+# ACR Docker login helper
 
+This is a wrapper for Docker Credential helpers created by Azure Container Registry (ACR) team. This credential helper make use of Azure Active Directory (AAD) to obtain and maintain user's credentials.
```
Comment (DavidObando) [inline README.md]: Please change to something like:

The ACR Docker Credential Helper allows users to sign-in to the Azure Container Registry service using their Azure Active Directory (AAD) credentials. This credential helper is in charge of ensuring that the stored credentials are valid, and when required it also renews the credentials for a repository.

For now, this credential helper works in tandem with the Azure CLI, which is required in order to initiate the credential flow. Once you've successfully logged in to your container registry with the Azure CLI, the credential helper administers the life cycle of your locally stored credential.

File: README.md [hunk 2]
```diff
@@ -1,2 +1,28 @@
...
+
+## Prerequisites
+
+- Docker installation is required, of course.
```
Comment (DavidObando) [inline README.md]: Change to:
```
- [Docker](https://www.docker.com/)
```

File: README.md [hunk 3]
```diff
@@ -1,2 +1,28 @@
...
+- [Azure CLI](https://github.com/Azure/azure-cli)
+
+## Installation
+For windows run the [powershell installation script](https://aka.ms/acr/installaad/win) in administrator mode:
+
+`iex ([System.Text.Encoding]::ASCII.GetString((Invoke-WebRequest -Uri https://aka.ms/acr/installaad/win).Content))`
```
Comment (DavidObando) [inline README.md]: Consider using `UTF8` instead.

File: README.md [hunk 4]
```diff
@@ -1,2 +1,28 @@
...
+
+For linux, run the [bash installation script](https://aka.ms/acr/installaad/bash) as root:
+
+`curl -L https://aka.ms/acr/installaad/bash | sudo /bin/bash`
+
+## Usage
+To login to an ACR service as followed:
+    `az acr login -n <registry name>`
+
+After that, you will be able to use docker normally. This credential helper would help maintaining your AAD access tokens.
```
Comment (DavidObando) [inline README.md]: Change the last part to: This credential helper will help maintaining your credentials valid.

File: README.md [hunk 5]
```diff
@@ -1,2 +1,28 @@
...
+
+- Docker installation is required, of course.
+- [Azure CLI](https://github.com/Azure/azure-cli)
+
+## Installation
+For windows run the [powershell installation script](https://aka.ms/acr/installaad/win) in administrator mode:
+
+`iex ([System.Text.Encoding]::ASCII.GetString((Invoke-WebRequest -Uri https://aka.ms/acr/installaad/win).Content))`
+
+For linux, run the [bash installation script](https://aka.ms/acr/installaad/bash) as root:
```
Comment (DavidObando) [inline README.md]: For Linux and macOS,

File: README.md [hunk 6]
```diff
@@ -1,2 +1,28 @@
...
+
+`curl -L https://aka.ms/acr/installaad/bash | sudo /bin/bash`
+
+## Usage
+To login to an ACR service as followed:
```
Comment (DavidObando) [inline README.md]: Change to:
After installing the ACR Docker Credential Helper, login to an Azure Container Registry using the Azure CLI:

File: README.md [hunk 7]
```diff
@@ -1,2 +1,28 @@
...
+    `az acr login -n <registry name>`
+
+After that, you will be able to use docker normally. This credential helper would help maintaining your AAD access tokens.
+
+## Troubleshooting
+- Why am I getting 401 error (authentication required)?
+
+    If you have not called `az acr login -n <registry>` to log in to your registry for extended period of time, please re-login with that command. If you find yourself having to re-login every hour or so, make sure your system clock time is correct.
```
Comment (DavidObando) [inline README.md]: If you have not called `az acr login -n <registry>` to log in to your registry for an extended period of time and you get a 401 error, please log in again. If you find yourself having to log in every hour or so, make sure your computer clock is set to the correct time.

File: README.md [hunk 8]
```diff
@@ -1,2 +1,28 @@
...
+`curl -L https://aka.ms/acr/installaad/bash | sudo /bin/bash`
+
+## Usage
+To login to an ACR service as followed:
+    `az acr login -n <registry name>`
+
+After that, you will be able to use docker normally. This credential helper would help maintaining your AAD access tokens.
+
+## Troubleshooting
+- Why am I getting 401 error (authentication required)?
```
Comment (DavidObando) [inline README.md]: Change for:
```
### Getting 401 (authentication required)
```

File: install/install.ps1 [hunk 9]
```diff
@@ -10,25 +10,25 @@ if ($systemStr -match '(x64)') {
...
 
 if (!$isAdmin) {
     Write-Error "Please run this script as administrator"
-    exit -1
+    break
 }
 
 $dockerLocation = $(where.exe docker)
 if (!$dockerLocation) {
     Write-Error "Please install docker first"
```
Comment (DavidObando) [inline install/install.ps1]: Change to:
Write-Error "Cannot find docker in path. Ensure it's installed and that its path is accessible."

File: install/install.sh [hunk 10]
```diff
@@ -1,5 +1,5 @@
 #!/bin/bash
-baseurl="https://github.com/Azure/acr-docker-credential-helper/releases/download/{TRAVIS_TAG}"
+baseurl="https://aadacr.blob.core.windows.net/acr-docker-credential-helper"
 unamestr=`uname`
 
 if [[ "$unamestr" == "Linux" ]]; then
```
Comment (DavidObando) [inline install/install.sh]: This is not needed. Given the updates to the build scripts, we'll download from `baseurl` for the specific OS (Linux or Darwin). Let's sync about this in the morning.

**Review by DavidObando — CHANGES_REQUESTED** [review]
A few changes are needed.

File: README.md [hunk 11]
```diff
@@ -1,2 +1,30 @@
...
+
+- [Docker](https://www.docker.com/)
+- [Azure CLI](https://github.com/Azure/azure-cli)
+
+## Installation
+For Windows, run the [powershell installation script](https://aka.ms/acr/installaad/win) in administrator mode:
+
+`iex ([System.Text.Encoding]::UTF8.GetString((Invoke-WebRequest -Uri https://aka.ms/acr/installaad/win).Content))`
+
+For Linux and MacOS, run the [bash installation script](https://aka.ms/acr/installaad/bash) as root:
```
Comment (DavidObando) [inline README.md]: macOS should have a lowercase m.

File: build/build-config-edit.sh [hunk 12]
```diff
@@ -22,7 +23,7 @@ do
     else
         exe_extension=""
     fi
-    outputFile="${bindir}/config-edit-${go_os}${exe_extension}"
+    outputFile="${bindir}/${go_os}/config-edit-${go_os}${exe_extension}"
```
Comment (DavidObando) [inline build/build-config-edit.sh]: We already have `$go_os` in the filename, why also in the path?

File: build/package.sh [hunk 13]
```diff
@@ -1,30 +1,37 @@
...
 fi
 
-pkgFile="docker-credential-acr-${osname}-amd64.tar.gz"
-pushd ${bindir}
-tar czf ${pkgFile} *
-popd
-mv ${bindir}/$pkgFile ${artifactsdir}
-
-echo "Packaged in ${pkgFile}"
+for osname in $(ls $binroot) ; do
```
Comment (DavidObando) [inline build/package.sh:20]: I see, so you used `$go_os` in the path above so you could do this. I recommend you simply loop around "linux", "darwin" and "windows".

File: build/package.sh [hunk 14]
```diff
@@ -1,30 +1,37 @@
...
+    bindir="$binroot/$osname"
+    if [[ -d $bindir ]]; then
+        echo "Packaging for ${osname}"
+        pkgName="docker-credential-acr-${osname}-amd64"
```
Comment (DavidObando) [inline build/package.sh]: Set "amd64" in a variable and append the variable here. We may build this helper for other architectures in the future and we want to have this ready to go.

File: src/docker-credential-acr/acr_login.go [hunk 15]
```diff
@@ -71,7 +71,7 @@ func GetUsernamePassword(serverAddress string, identityToken string) (user strin
 		var challenge *authDirective
 		if challenge, err = receiveChallengeFromLoginServer(serverAddress); err != nil {
 			// ignore all error when receiving the challenge
-			logrus.Warnf("[Azure Login Helper] server %s does not responded with a valid challenge, reverting to default login...\nerror: %s\n", serverAddress, err)
+			logrus.Infof("[Azure Login Helper] server %s does not responded with a valid challenge, reverting to default login...\nerror: %s\n", serverAddress, err)
```
Comment (DavidObando) [inline src/docker-credential-acr/acr_login.go]: Typo, change to:
```go
logrus.Infof("[Azure Login Helper] server %s didn't respond with a valid challenge, reverting to default login...\nerror: %s\n", serverAddress, err) 
```

File: build/package.sh (hunk 13 above)
Comment (shhsu) [reply build/package.sh:20]: Same as above

File: build/build-config-edit.sh (hunk 12 above)
Comment (shhsu) [reply build/build-config-edit.sh]: It makes the bin folder more neat. Right now all binaries are in the same directory.

**Review by DavidObando — APPROVED** [review]

File: src/docker-credential-acr/acr_login.go (hunk 15 above)
Comment (shhsu) [reply src/docker-credential-acr/acr_login.go]: FYI, I have now fixed the typo (see file changed tab) but somehow the PR comment section didn't show it's fixed.

**Pull request merged**